ProjFS.Mac/PrjFSKextTests/MessageTests.mm

```diff
@@ -0,0 +1,44 @@
+#import <XCTest/XCTest.h>
+
+#include "../PrjFSKext/Message_Kernel.hpp"
+
+@interface MessageTests : XCTestCase
+
+@end
+
+@implementation MessageTests
+
+- (void)setUp
+{
+}
+
+- (void)tearDown
+{
+}
+
+- (void)testMessageEncodedSize
+{
+    MessageHeader testHeader = {};
+    XCTAssertEqual(sizeof(testHeader), Message_EncodedSize(&testHeader));
+    
+    testHeader.pathSizesBytes[MessagePath_Target] = UINT16_MAX;
+    testHeader.pathSizesBytes[MessagePath_From] = UINT16_MAX;
+    
+    static const uint64_t maxSize = sizeof(testHeader) + UINT16_MAX * 2llu;
+    XCTAssertEqual(maxSize, Message_EncodedSize(&testHeader));
+    
+    srand(0);
+    for (unsigned i = 0; i < 10000; ++i)
+    {
+        uint16_t targetSize = testHeader.pathSizesBytes[MessagePath_Target] = rand() & UINT16_MAX;
+        uint16_t fromSize = testHeader.pathSizesBytes[MessagePath_From] = rand() & UINT16_MAX;
+        uint32_t size = Message_EncodedSize(&testHeader);
+        XCTAssertGreaterThanOrEqual(size, sizeof(testHeader));
+        XCTAssertLessThanOrEqual(size, maxSize);
+        XCTAssertGreaterThan(size, targetSize);
+        XCTAssertGreaterThan(size, fromSize);
```

Comment:
Can we also have an `XCTAssertEqual` for `size` (or perhaps in place of the other checks)?

```
XCTAssertEqual(size, sizeof(testHeader) + testHeader.pathSizesBytes[MessagePath_Target] + testHeader.pathSizesBytes[MessagePath_From]);
```

Reply:
The question then becomes what exactly we are testing. The equals test is something of a tautological `XCTAssert(a + b == a + b);` kind of test where the implementation under test is the same as the test itself, which doesn't really say much.

What I'm trying to do with the inequality checks is testing various properties of the result that must not be violated, without actually assuming a specific implementation.

Reply:
> What I'm trying to do with the inequality checks is testing various properties of the result that must not be violated, without actually assuming a specific implementation.

In that case would it be fair to also check:

```
XCTAssertGreaterThanOrEqual(size, sizeof(testHeader) + targetSize + fromSize);
```

That would ensure the encoded size is allowing enough room for the header and the paths.